### DISCUSSION TITLE
Database queries in getStaticProps (tutorial question)

### DISCUSSION THREAD

**Discussion by ohthepain** [vercel/next.js#55230] [Help] [answered]
### Summary

In [this tutorial](https://nextjs.org/learn/basics/data-fetching/getstaticprops-details) it says that 

> In production, [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) runs at build time.

but also

> You can also query the database directly:

Why would I query the database at build time?

**Comment by icyJoseph** [comment, marked as answer]
The idea is to pre-render and have a statically generated page during build time. This is most useful for data that doesn't change as often.

For data that changes relatively often you can combine this with ISR (return revalidate set to a given number), and that'll re-generate the page during runtime.

Take for instance, a book's data. It'll hardly ever change, so you can query your DB for a given book data, and return that as part of `getStaticProps`(GSP), and then Next.js will pre-render an actual HTML page with that books UI, as well as JavaScript to hydrate the UI client side.

Now imagine you have tons of books. You could pre-render a few at build time, say the top 5% most purchased. Again you get this data at build time. This time you can introduce `getStaticPaths`, with fallback set to `true` or `'blocking'`, and that'll allow rendering other books that were not part of that 5%, during runtime. Typically when a Link points to the book page, or when a user lands on a non-pre-rendered page.

Data that's "hot"/"live", instantly changing, etc, you'd probably be better of getting it from the client side, or through `getServerSideProps` if SEO is a concern.

**Reply by ohthepain** [reply]
Cool use case, thank you.
